program/src/processor.rs

```diff
@@ -86,8 +100,18 @@ fn relocate_lamports(
 // constructs it the unchecked mode doesnt add pubkeys of non-signers, relying
 // on downstream errors if a required signer is missing the checked mode expects
 // every AccountInfo passed in should be a signer and errors if any is not
+// JC: I kept having to re-read what this was doing to be sure it was ok in
+// every context that it's used.
+// I appreciate that this is more work for you, but could you add a comment for
+// every single time it's called with `checked` as `false` and explain where the
+// signer will be checked? Ie. "unchecked because meta.authorize() will check
+// that the stake or withdraw authority signed".
+// It would certainly make me feel better and help ensure that we preserve all
+// signer checks.
```

Comment:
this function confuses me too, ill think about a better interface, even if its just replacing the bool with a checked `collect_signers` and a `collect_signers_unchecked`... yea actually im just going to rethink this. i want to check signers when we make the hashmap, instead of worrying about checking them explicitly in every processor and potentially missing one, so its correct by construction. and i did it right but the interface i chose is not good

the problem with `collect_signers` is its serving two masters. in the existing stake program, a _lot_ of process functions dont check signers and rely on `Authorized` to do the checks internally. for those, i check the signers myself, and added `NOTE` comments where i did it differently in case i later realize through testing that i broke compat by being too strict

however `collect_signers` is _also_ used for three unchecked processors (set lockup, authorize, authorize with seed) that have checked variants, where the checked ones require that new authorities sign as a safety feature, while the unchecked ones just take a pubkey. but because of the nature of the `Authorized` functions these two kinds of things get mixed together in the same hashmap. its also confusing because they all take different arguments structs, i even had to go back and look at `sdk/program/src/stake/instruction.rs` to remind myself how these work

ok yea i think i tried to make this too modular, im just going to have two different functions that behave differently to make it make more sense